### PR TITLE
site_alerts: add runtime validation for critical.externalURL

### PR DESCRIPTION
**Pull request by unknwon**
This PR adds runtime validation for configuration apart from using `conf.ContributeValidator` because latter will prevent server from starting. Only `externalURL` is validated at this point.

This fixes the original problem reported in https://github.com/sourcegraph/sourcegraph/issues/3899.

**Test plan**: Comment out `externalURL` in `dev-private` and see if the site alert shows up. And disappear after set a non-empty value via management console.

**UI preview**:
image

### CODE REVIEW

**Comment by unknwon** [comment]
I think the use of "runtime" is not very accurate, another short word has meaning of "after initialization time" would be better.

File: cmd/frontend/graphqlbackend/site_alerts.go [hunk 1]
```diff
@@ -79,6 +93,18 @@ func init() {
 				},
 			}
 		}
+
+		runtimeProblems, err := doRuntimeValidation()
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/site_alerts.go]:
```suggestion
		configWarnings, err := getConfigWarnings()
```

File: cmd/frontend/graphqlbackend/site_alerts.go [hunk 2]
```diff
@@ -79,6 +93,18 @@ func init() {
...
+		if err != nil {
+			return []*Alert{
+				{
+					TypeValue:    AlertTypeError,
+					MessageValue: `Update [**critical configuration**](/help/admin/management_console) to resolve problems: ` + err.Error(),
+				},
+			}
+		}
+		problems = append(problems, runtimeProblems...)
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/site_alerts.go]:
```suggestion
		problems = append(problems, configWarnings...)
```

File: cmd/frontend/graphqlbackend/site_alerts.go [hunk 3]
```diff
@@ -60,6 +61,19 @@ func (r *siteResolver) Alerts(ctx context.Context) ([]*Alert, error) {
 	return alerts, nil
 }
 
+// doRuntimeValidation performs additional validations at runtime which otherwise will
+// prevent server from starting via conf.ContributeValidator.
+func doRuntimeValidation() (problems conf.Problems, err error) {
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/site_alerts.go]:
```suggestion
// getConfigWarnings identifies problems with the configuration that a site
// admin should address, but do not prevent Sourcegraph from running.
func getConfigWarnings() (problems conf.Problems, err error) {
```

**Comment by slimsag** [comment]
Probably a better solution would be to have `conf.Validate` tell if the problems it returns are fatal or not (should prevent server from starting), e.g. through a `warning bool` field. But this is good enough for now and simple :)

File: cmd/frontend/graphqlbackend/site_alerts.go [hunk 4]
```diff
@@ -60,6 +61,21 @@ func (r *siteResolver) Alerts(ctx context.Context) ([]*Alert, error) {
 	return alerts, nil
 }
 
+// doRuntimeValidation performs additional validations at runtime which otherwise will
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/site_alerts.go]:
```suggestion
```

File: cmd/frontend/graphqlbackend/site_alerts.go [hunk 5]
```diff
@@ -60,6 +61,21 @@ func (r *siteResolver) Alerts(ctx context.Context) ([]*Alert, error) {
...
+// prevent server from starting via conf.ContributeValidator.
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/site_alerts.go]:
```suggestion
```

**Pull request merged**